### PR TITLE
louvain- adding edge projection

### PR DESCRIPTION
**Pull request by uditagar**
THis is an initial attempt to support edge projection.

### CODE REVIEW

**Comment by uditagar** [comment]
I have throughly testd this version now on GAP-road with different edge projections.

**Comment by gurbinder533** [comment]
Can you please fix the formatting and other issues so that all the CI tests pass?

**Comment by uditagar** [comment]
> Can you please fix the formatting and other issues so that all the CI tests pass?

I am not able to fix it. Because ctest graph has only one edge property, whereas this app requires two edge properties. I couldn't figure how to resolve it. Let me know how to do that. thanks

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h [hunk 1]
```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
     return katana::ResultSuccess();
   }
 
+  template <typename NodeData, typename EdgeData, typename EdgeWeightType>
+  katana::Result<std::unique_ptr<katana::PropertyGraph>> GraphFiltering(
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/ClusteringImplementationBase.h]: Call it `ProjectGraph`

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h [hunk 2]
```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
...
+      const Graph& graph, katana::PropertyGraph* pfg_mutable,
+      const std::vector<std::string>& temp_node_property_names,
+      const std::vector<std::string>& temp_edge_property_names) {
+    //	  using GNode = typename Graph::Node;
+
+    std::vector<std::vector<uint32_t>> edges_id(graph.num_nodes());
+    std::vector<std::vector<std::tuple<EdgeWeightType, uint8_t>>> edges_data(
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/ClusteringImplementationBase.h]: `katana::LargeArray<gstl::Vector<...` for both

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h [hunk 3]
```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
...
+                  graph.template GetEdgeData<EdgeWeight<EdgeWeightType>>(ii),
+                  1));
+            }
+          }
+        },
+        katana::steal());
+
+    uint64_t num_nodes_next = graph.num_nodes();
+
+    /* Serial loop to reduce all the edge counts */
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/ClusteringImplementationBase.h]: Drop "serial loop to"

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h [hunk 4]
```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
...
+    std::vector<uint64_t> prefix_edges_count(graph.num_nodes());
+    katana::GAccumulator<uint64_t> num_edges_acc;
+    katana::do_all(
+        katana::iterate(uint64_t{0}, graph.num_nodes()), [&](uint64_t c) {
+          prefix_edges_count[c] = edges_id[c].size();
+          num_edges_acc += prefix_edges_count[c];
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/ClusteringImplementationBase.h:532]: Both of these can be computed in the previous `do_all`.

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h [hunk 5]
```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
...
+                  1));
+            }
+          }
+        },
+        katana::steal());
+
+    uint64_t num_nodes_next = graph.num_nodes();
+
+    /* Serial loop to reduce all the edge counts */
+    std::vector<uint64_t> prefix_edges_count(graph.num_nodes());
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/ClusteringImplementationBase.h]: `katana::LargeArray`

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h [hunk 6]
```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
...
+    katana::GAccumulator<uint64_t> num_edges_acc;
+    katana::do_all(
+        katana::iterate(uint64_t{0}, graph.num_nodes()), [&](uint64_t c) {
+          prefix_edges_count[c] = edges_id[c].size();
+          num_edges_acc += prefix_edges_count[c];
+        });
+
+    uint64_t num_edges_next = num_edges_acc.reduce();
+    for (uint64_t c = 1; c < graph.num_nodes(); ++c) {
+      prefix_edges_count[c] += prefix_edges_count[c - 1];
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/ClusteringImplementationBase.h:540]: You can two `on_each` to do this in parallel.

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h [hunk 7]
```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
...
+        prefix_edges_count[graph.num_nodes() - 1] == num_edges_next);
+
+    // Remove all the existing node/edge properties
+    for (auto property : temp_node_property_names) {
+      if (auto r = pfg_mutable->RemoveNodeProperty(property); !r) {
+        return r.error();
+      }
+    }
+    for (auto property : temp_edge_property_names) {
+      if (auto r = pfg_mutable->RemoveEdgeProperty(property); !r) {
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/ClusteringImplementationBase.h]: Why do you need to remove properties?

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h [hunk 8]
```diff
@@ -518,6 +520,150 @@ struct ClusteringImplementationBase {
...
+    katana::do_all(
+        katana::iterate(uint64_t{0}, num_edges_next),
+        [&](uint64_t i) { out_dests_next[i] = topology.out_dests->Value(i); },
+        katana::no_stats());
+
+    auto topo_next = std::make_unique<katana::GraphTopology>(
+        std::move(out_indices_next), std::move(out_dests_next));
+
+    auto pfg_next = std::make_unique<katana::PropertyGraph>();
+    if (auto r = pfg_next->SetTopology(std::move(topo_next)); !r) {
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/ClusteringImplementationBase.h:572]: Return a new `PropertyGraph`. Don't modify the existing `PropertyGraph`.

File: libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp [hunk 9]
```diff
@@ -382,18 +383,46 @@ struct LouvainClusteringImplementation
...
     if (!graph_result) {
       return graph_result.error();
     }
     Graph graph_curr = graph_result.value();
 
+    if (!edge_type_property_name.empty()) {
+      katana::do_all(
+          katana::iterate((uint64_t)0, graph_curr.num_nodes()),
+          [&](uint64_t n) {
+            graph_curr.template GetData<CurrentCommunityId>(n) = n;
```
Comment (roshandathathri) [inline libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp]: This should be done after edge projection

**Review by roshandathathri — CHANGES_REQUESTED** [review]

File: libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp (hunk 9 above)
Comment (uditagar) [reply libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp]: I am using this for graph filtering. They are re-initialied anyways at the start of a  new coarsened graph iteration. thanks

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h (hunk 8 above)
Comment (uditagar) [reply libgalois/include/katana/analytics/ClusteringImplementationBase.h:572]: In this code, I am modifying the duplicated PropertyGraph. Original graph is not passed to this fucntion. thanks

Comment (roshandathathri) [reply libgalois/include/katana/analytics/ClusteringImplementationBase.h:572]: Why do you need to duplicate? Create a new graph with the required topology

File: libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp (hunk 9 above)
Comment (roshandathathri) [reply libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp]: Why do you need this for filtering?

Comment (uditagar) [reply libgalois/src/analytics/louvain_clustering/louvain_clustering.cpp]: got it now, thanks

File: libgalois/include/katana/analytics/ClusteringImplementationBase.h (hunk 8 above)
Comment (uditagar) [reply libgalois/include/katana/analytics/ClusteringImplementationBase.h:572]: In the original louvain, the duplicate graph is the one that gets coarsened, while the original is left untouched for verification purposes later on. Here I am taking that duplicated graph and then filtering out edges from it

**Comment by uditagar** [comment]
closing this since we now have a better way to support projections. thanks

**Pull request closed without merging**